### PR TITLE
impl(rest): improve write performance

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9543)

### CODE REVIEW

File: google/cloud/internal/curl_impl.cc [hunk 1]
```diff
@@ -528,49 +532,58 @@ StatusOr<int> CurlImpl::PerformWork() {
...
                     << ", multi_remove_status=" << multi_remove_status << "\n";
 
       // Ignore errors when closing the handle. They are expected because
       // libcurl may have received a block of data, but the WriteCallback()
       // (see above) tells libcurl that it cannot receive more data.
       if (closing_) continue;
-      if (!status.ok()) return status;
-      if (!multi_remove_status.ok()) return multi_remove_status;
+      if (result != CURLM_OK)
+        return CurlHandle::AsStatus(msg->data.result, __func__);
```
Comment (coryan) [inline google/cloud/internal/curl_impl.cc:551]: nit: braces, we allow no braces only if it all fits in a line.

File: google/cloud/internal/curl_impl.cc [hunk 2]
```diff
@@ -528,49 +532,58 @@ StatusOr<int> CurlImpl::PerformWork() {
...
+      if (!multi_remove_status.ok())
+        return AsStatus(multi_remove_result, __func__);
```
Comment (coryan) [inline google/cloud/internal/curl_impl.cc:553]: more missing braces

File: google/cloud/internal/curl_impl.cc [hunk 3]
```diff
@@ -763,6 +776,10 @@ Status CurlImpl::MakeRequest(CurlImpl::HttpMethod method,
     if (!status.ok()) return OnTransferError(std::move(status));
     status = handle_.SetOption(CURLOPT_UPLOAD, 1L);
     if (!status.ok()) return OnTransferError(std::move(status));
+    status = handle_.SetOption(CURLOPT_INFILESIZE_LARGE,
+                               static_cast<long>(writev.size()));  // NOLINT
```
Comment (coryan) [inline google/cloud/internal/curl_impl.cc:780]: I think this is the number of elements in the vector, not the total number of bytes?  In other words, you are calling (for example) `handle_.SetOption(CURLOPT_INFILESIZE_LARGE, 2)` when you want `handle_.SetOption(CURLOPT_INFILESIZE_LARGE, writev[0].size() + writev[1].size())` ?

**Review by coryan — CHANGES_REQUESTED** [review]

File: google/cloud/internal/curl_impl.cc (hunk 3 above)
Comment (devbww) [reply google/cloud/internal/curl_impl.cc:780]: I had wondered the same thing, but it appears that `WriteVector.size()` is, perhaps suprisingly, not the size of the vector but rather the sum of the sizes over the vector, as you desire.  (In contrast, however, `WriteVector.empty()` is just on the vector.)

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9543)* status: 0 of 1 files reviewed, 3 unresolved discussions (waiting on @coryan and @devbww)

___
*[`google/cloud/internal/curl_impl.cc` line 551 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9543#-N7T6guZVf9jWOPHteMz-r2-551:-N7WeumdC2gP4VmXcVuR:b-6i7zej) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/db54247088b80f113c78440f29417c1501c4da51/google/cloud/internal/curl_impl.cc#L551)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: braces, we allow no braces only if it all fits in a line.
</blockquote></details>

I wish clang-format would add them as it splits lines...
___
*[`google/cloud/internal/curl_impl.cc` line 553 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9543#-N7T6guZVf9jWOPHteMz-r2-553:-N7Weztv2hBdsZ8Foq4I:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/db54247088b80f113c78440f29417c1501c4da51/google/cloud/internal/curl_impl.cc#L553)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

more missing braces
</blockquote></details>

Done.
___
*[`google/cloud/internal/curl_impl.cc` line 780 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9543#-N7T6guZVf9jWOPHteMz-r2-780:-N7Wf1129PLbWlNqqkDQ:b-q6n0ed) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/db54247088b80f113c78440f29417c1501c4da51/google/cloud/internal/curl_impl.cc#L780)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

I had wondered the same thing, but it appears that `WriteVector.size()` is, perhaps suprisingly, not the size of the vector but rather the sum of the sizes over the vector, as you desire.  (In contrast, however, `WriteVector.empty()` is just on the vector.)
</blockquote></details>

My rationale is that `WriteVector` is supposed to make n number of spans appear contiguous. It shouldn't matter how many spans are contained. Thus `size` gives you the total bytes across all spans and `empty` indicates whether there are 0 bytes or more than 0 bytes present, regardless of the number of spans.

**Review by devbww — APPROVED** [review]
> Reviewed 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9543)* status: all files reviewed, 3 unresolved discussions (waiting on @coryan)

___
*[`google/cloud/internal/curl_impl.cc` line 780 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9543#-N7T6guZVf9jWOPHteMz-r2-780:-N7WgHjUDck4jAwrLo7b:b-uz2hp6) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/db54247088b80f113c78440f29417c1501c4da51/google/cloud/internal/curl_impl.cc#L780)):*
> `empty` indicates whether there are 0 bytes or more than 0 bytes present, regardless of the number of spans

Not if there were spans, but they were all empty (i.e. `size() == 0` does not currently imply `empty()`).

**Review by devbww — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9543)* status: all files reviewed, 3 unresolved discussions (waiting on @coryan)

___
*[`google/cloud/internal/curl_impl.cc` line 780 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9543#-N7T6guZVf9jWOPHteMz-r2-780:-N7WhmrH54kCNcQTH8x0:b-hliedb) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/db54247088b80f113c78440f29417c1501c4da51/google/cloud/internal/curl_impl.cc#L780)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

> `empty` indicates whether there are 0 bytes or more than 0 bytes present, regardless of the number of spans

Not if there were spans, but they were all empty (i.e. `size() == 0` does not currently imply `empty()`).
</blockquote></details>

[Sorry ... didn't mean to approve underneath @coryan]

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9543#-:-N7Wm5eBCcGefFeR5Hfz:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @scotthart)

___
*[`google/cloud/internal/curl_impl.cc` line 780 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9543#-N7T6guZVf9jWOPHteMz-r2-780:-N7WltnZ0cJlJ0ov6gTE:b-ts9fl) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/db54247088b80f113c78440f29417c1501c4da51/google/cloud/internal/curl_impl.cc#L780)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

[Sorry ... didn't mean to approve underneath @coryan]
</blockquote></details>

Ah, I thought it was just an alias.  I should have checked.  We can improve the API for `WriteVector` separately.  @scotthart consider adding a bug to remind us.

**Pull request merged**